### PR TITLE
PR: Store autosave information in separate files instead of the Spyder config

### LINKED ISSUE #9420: FileNotFoundError when closing a file that's opened in several Spyder instances
### Issue Report Checklist

* [ ] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ ] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
Error when closing a tab whose autosave file was destroyed.

### What steps reproduce the problem?

1. 
2. 
3. 

### Paste Traceback/Error Below (if applicable)

```python-traceback
ERROR:spyder.plugins.editor.widgets.autosaveerror:<class 'FileNotFoundError'> while removing autosave file /Users/quentinpeter/.spyder-py3-dev/autosave/spydercommbase.py
Traceback (most recent call last):
  File "/Users/quentinpeter/pyscripts/spyder/spyder/plugins/editor/utils/autosave.py", line 165, in remove_autosave_file
    os.remove(autosave_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/Users/quentinpeter/.spyder-py3-dev/autosave/spydercommbase.py'
```

## Versions

* Spyder version: 
* Python version: 
* Qt version: 
* PyQt version: 
* Operating System name/version: 

### Dependencies

```
PASTE DEPENDENCIES HERE
```

### PR DESCRIPTION
**Pull request by jitseniesen**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/)) **N/A** 

This PR stores which files are autosaved and where in a separate file called `pidNNN.txt` in the autosave directory, where `NNN` is the pid of the Spyder process. This information is used when the files are recovered after a crash. Before this information was stored in the Spyder config file.

The new approach has two advantages. Firstly, if the user is running multiple Spyder instances simultaneously, then one instance overwrites the changes the other instance makes to the config file, so the autosave info got muddled in the old system, while now each instance has its own location to write its data to. Secondly, by also recording the pid of the Spyder process that created the autosave file, we can check whether that process is still running. and not change the autosave file if it is.

### Issue(s) Resolved

Fixes #9420

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
Jitse Niesen

### CODE REVIEW

File: spyder/utils/programs.py [hunk 1]
```diff
@@ -959,3 +962,30 @@ def check_python_help(filename):
...
+        names = set(['spyder', 'spyder3', 'spyder.exe', 'spyder3.exe',
+                     'bootstrap.py', 'spyder-script.py'])
+        if running_under_pytest():
+            names.add('runtests.py')
+
+        # Check the first three command line arguments
+        arguments = set(os.path.basename(arg) for arg in p.cmdline()[:3])
+        conditions = [names & arguments]
+        return any(conditions)
+
```
Comment (ccordoba12) [inline spyder/utils/programs.py]: Please remove this blank line.

File: spyder/plugins/editor/utils/autosave.py [hunk 2]
```diff
@@ -106,33 +109,78 @@ def stop_autosave_timer(self):
...
+                                        in txt_as_dict.items()]
+                pid = int(match.group(1))
+                if is_spyder_process(pid):
+                    logger.debug('Ignoring files in {}'.format(full_name))
+                else:
+                    files_to_recover += list(txt_as_dict.items())
+            else:
+                non_pid_files.append(full_name)
+
+        # Add all files not mentioned in any pid file
```
Comment (ccordoba12) [inline spyder/plugins/editor/utils/autosave.py]: When does this case happen?

File: spyder/plugins/editor/utils/autosave.py [hunk 3]
```diff
@@ -106,33 +109,78 @@ def stop_autosave_timer(self):
...
+        the pid files.
+        """
+        files_to_recover, pidfiles = self.get_files_to_recover()
+        dialog = RecoveryDialog(files_to_recover, parent=self.editor)
         dialog.exec_if_nonempty()
         self.recover_files_to_open = dialog.files_to_open[:]
+        for pidfile in pidfiles:
+            try:
+                os.remove(pidfile)
+            except OSError:
```
Comment (ccordoba12) [inline spyder/plugins/editor/utils/autosave.py]: Perhaps we should also incluce IOError here?

File: spyder/plugins/editor/utils/autosave.py [hunk 4]
```diff
@@ -192,6 +240,26 @@ def create_unique_autosave_filename(self, filename, autosave_dir):
...
+        autosave_dir = get_conf_path('autosave')
+        my_pid = os.getpid()
+        pidfile_name = osp.join(autosave_dir, 'pid{}.txt'.format(my_pid))
+        if self.name_mapping:
+            with open(pidfile_name, 'w') as pidfile:
+                pidfile.write(pprint.pformat(self.name_mapping))
+        else:
+            try:
+                os.remove(pidfile_name)
+            except OSError:
```
Comment (ccordoba12) [inline spyder/plugins/editor/utils/autosave.py]: Perhaps we should also incluce IOError here?

File: spyder/plugins/editor/utils/autosave.py (hunk 2 above)
Comment (jitseniesen) [reply spyder/plugins/editor/utils/autosave.py]: It shouldn't, but may happen if the pid file got corrupted. I expanded the comment to clarify.

File: spyder/plugins/editor/utils/autosave.py (hunk 3 above)
Comment (jitseniesen) [reply spyder/plugins/editor/utils/autosave.py]: Definitely. I will be glad when Python 2 disappears!

**Comment by jitseniesen** [comment]
> I left a very small review because I trust your work to solve this problem.

Thanks, though I'm sure that it's possible to find more things that can be improved.

**Pull request merged**